Add support for highlighting chat messages

frenzibyte opened the pull request:
Closes #17088 

https://user-images.githubusercontent.com/22781491/156957613-c4485659-7bfc-4418-8952-2205eee91075.mp4

I haven't thoroughly tested how this behaves yet in a real client, but I'm mostly PR'ing this for feedback on the direction and specifically the logic in chat overlay to handle highlighting messages on non-loaded `DrawableChannel`s (via `messagePendingHighlight` and `tryHighlightMessage`).

- [x] Depends on #17115 
- [x] Depends on ppy/osu-framework#5058
- [x] Depends on ppy/osu-framework#5059 (see [this](https://github.com/ppy/osu/blob/30b38345aad33a37419f32528cc1ff831e70699a/osu.Game/Overlays/Chat/DrawableChannel.cs#L91) and [this](https://github.com/ppy/osu/blob/30b38345aad33a37419f32528cc1ff831e70699a/osu.Game/Overlays/Chat/DrawableChannel.cs#L121))
- [x] Depends on framework bump
- [x] Pending actual game testing

frenzibyte committed:
Add ability to highlight chat lines

frenzibyte committed:
Add test coverage for channel message highlighting

frenzibyte committed:
Add support to highlight messages in chat overlay

frenzibyte committed:
Add test coverage for chat overlay message highlighting

frenzibyte committed:
Hook up message notifications to chat message highlighting logic

peppy commented:
I like the effect, it should work fine visually for now.

bdach commented on osu.Game/Overlays/Chat/ChatLine.cs:
```diff
@@ -200,13 +213,21 @@ protected override void LoadComplete()
+        /// <remarks>
+        /// Scheduling is required to ensure the animation doesn't play until the chat line is in view and not scrolled away.
+        /// </remarks>
+        public void ScheduleHighlight() => Schedule(() => lineBackground.FlashColour(usernameColour.Darken(1f).Opacity(0.5f), 1000, Easing.InQuint));
```
Nitpick but technically the more proper way to do this fade would be to do something the lines of

```suggestion
        public void ScheduleHighlight() => Schedule(() =>
        {
            var backgroundColour = usernameColour.Darken(1f);
            lineBackground.Colour = backgroundColour.Opacity(0);
            lineBackground.FlashColour(backgroundColour.Opacity(0.5f), 1000, Easing.InQuint);
        });
```

The flash as written in this PR will interpolate from the darkened-transparent `usernameColour` to fully transparent black, which means not only the alpha will change, the colour will actually fade to black at the same time as well. Setting `usernameColour` to have all components other than alpha the same fixes this, as the colour part will not interpolate, just the alpha will.

In fact this could be vastly simplified by assigning `lineBackground.Colour = usernameColour` and only modulating alpha via `FadeTo()` rather than doing these colour manipulations like here. Unless the darkening was somehow explicitly intended?

frenzibyte replied:
Can simplify by assigning the username colour to `lineBackground` directly as you mentioned, didn't do that as I was initially using a constant highlight colour, where it doesn't make sense for `lineBackground` to have such colour assigned to it.

frenzibyte commented on osu.Game/Overlays/ChatOverlay.cs:
```diff
@@ -290,19 +296,54 @@ private void currentChannelChanged(ValueChangedEvent<Channel> e)
+        /// <param name="message">The message to highlight.</param>
+        public void HighlightMessage(Message message)
+        {
+            if (currentDrawableChannel?.Channel.Id == message.ChannelId)
```
Right, all can use `currentDrawableChannel` safely.

frenzibyte committed:
Check channel ID on message highlight using `currentDrawableChannel`

frenzibyte committed:
OpenChannelNotification -> HighlightMessageNotification

frenzibyte committed:
Refactor message highlighting logic to rely on a `Channel` data bindable

frenzibyte committed:
Fix `Message` equality not passing on equal references

frenzibyte commented on osu.Game/Overlays/ChatOverlay.cs:
```diff
@@ -290,19 +296,54 @@ private void currentChannelChanged(ValueChangedEvent<Channel> e)
+                    {
+                        tryHighlightMessage(messagePendingHighlight);
+                        messagePendingHighlight = null;
+                    }
```
Good idea, that simplified most of the highlighting logic. The only thing I'm unsure how to feel about is `Channel.HighlightedMessage` being set back to null once `DrawableChannel` receives the callback (to avoid highlighting again when, for example, reloading a channel by leaving and joining again), but I'll leave that to you as to whether you think it's fine and within what you have expected.

peppy replied:
I think I'd be more okay with the bindable flow if it was renamed to `LastHighlightedMessage` and was not set to `null`. Would that still work?

bdach replied:
I don't immediately see a reason why it wouldn't work personally.

frenzibyte committed:
Update `ChannelManager.CurrentChannel` directly to handle non-loaded chat scenario

`currentChannel` gets instantiated once the chat overlay is open, while
`HighlightMessage` could be called while the chat overlay has never been
open.

This will all be rewritten with the new chat overlay design anyways, so
should be fine for now.

frenzibyte committed:
Use `Equals` instead of reference equality operator

frenzibyte committed:
Simplify message highlighting transforms

frenzibyte committed:
Revert `newMessagesArrived` changes and always schedule highlight

In the case a message arrives and the chat overlay is hidden, clicking
on the mention notification will not work as the `HighlightedMessage`
bindable callback will execute before the scheduled `newMessagesArrived`
logic (which was hanging since the message arrived until the chat
overlay became open because of the notification).

Simplify things by always scheduling the `HighlightedMessage` bindable
callback.

peppy committed:
Merge branch 'master' into chat-mention-highlight

peppy committed:
Update framework

peppy committed:
Fix unconditional null in `Equals` implementation

peppy commented on osu.Game/Overlays/Chat/ChatLine.cs:
```diff
@@ -141,6 +146,15 @@ private void load(OsuColour colours)
 
             InternalChildren = new Drawable[]
             {
+                lineHighlightBackground = new Container
```
Not sure about this existing from the outset. Any reason it can't be added when the highlight actually occurs? Seems like unnecessary overhead.

peppy reviewed (changes requested):
Feels good though.

frenzibyte committed:
Schedule chat line highlight after children to handle non-loaded lines

frenzibyte committed:
Scroll chat line to channel center

We may eventually want that encapsulated within `ScrollIntoView`, as it
would also come in handy for `GameplayLeaderboard`.

frenzibyte committed:
Increase highlight delay to 1500ms

frenzibyte committed:
Add test case for highlighting while chat overlay is hidden

frenzibyte committed:
Instantiate highlight background container on animation

Also removes the necessity of scheduling as it actually never worked as
intended, `Scheduler` will still update even when the chat line is
masked away, and the animation will never be held anyways.

The new duration of the animation should be enough for long scrolls
either way.

frenzibyte committed:
Add too many messages to better test long scrolls

frenzibyte commented:
I've improved the scrolling logic to ensure the highlighted message display at the middle of chat rather than anywhere in view:

https://user-images.githubusercontent.com/22781491/157566893-28550563-d233-4480-98dd-8e942d5539d2.mp4

peppy committed:
Merge branch 'master' into chat-mention-highlight

peppy commented:
I've made a couple of minor tweaks, explained in the commit messages (although for some reason the commits aren't showing up yet..).

Looks good otherwise!

peppy commented on osu.Game/Overlays/Chat/DrawableChannel.cs:
```diff
@@ -79,6 +82,34 @@ private void load()
+        {
+            base.LoadComplete();
+
+            highlightedMessage = Channel.HighlightedMessage.GetBoundCopy();
+            highlightedMessage.BindValueChanged(m =>
```
I don't understand why this is a bindable flow rather than an event (ie. `RequestHighlight` similar to what we use in [other](https://github.com/ppy/osu/blob/266b4c7124fe4b578087afaaba46e18ec4ee0423/osu.Game/Overlays/Login/LoginForm.cs#L29) [places](https://github.com/ppy/osu/blob/2a1edcdf73de482ded7335b5ea2fdf2e607ba300/osu.Game/Screens/OnlinePlay/DrawableRoomPlaylistItem.cs#L57). The value is being nulled as soon as it is consumed anyway, so this either needs further explanation or should be an event flow instead.

bdach replied:
See previous review conversation: https://github.com/ppy/osu/pull/17137#discussion_r821087954

I found the way this was originally written on events pretty ugly because of the need to potentially handle channel switches at the same time as highlighting the message.

peppy commented on osu.Game/Overlays/ChatOverlay.cs:
```diff
@@ -303,6 +303,18 @@ private void currentChannelChanged(ValueChangedEvent<Channel> e)
+        public void HighlightMessage(Message message)
+        {
+            if (channelManager.CurrentChannel.Value.Id != message.ChannelId)
+                channelManager.CurrentChannel.Value = channelManager.JoinedChannels.Single(c => c.Id == message.ChannelId);
```
Not 100% sure about using `.Single` here rather than `.FirstOrDefault`. Most other usages of setting the current channel gracefully fallback rather than throwing. Seems like if a highlight fails we'd not want the game to crash.

frenzibyte replied:
Now that I think about it, this could potentially fail if the notification comes from a channel that has now been left.

May wanna re-join the channel in that case... will see.

frenzibyte committed:
Add functionality to switch to successfully joined channel

frenzibyte committed:
Handle non-existing channels on message highlighting gracefully

frenzibyte committed:
Add test coverage for highlighting message on left channel

bdach commented on osu.Game/Online/Chat/ChannelManager.cs:
```diff
@@ -420,10 +420,11 @@ private Channel getChannel(Channel lookup, bool addToAvailable = false, bool add
+        /// <param name="switchToJoined">Switch <see cref="CurrentChannel"/> to the joined channel on successful request.</param>
         /// <returns>The joined channel. Note that this may not match the parameter channel as it is a backed object.</returns>
-        public Channel JoinChannel(Channel channel) => joinChannel(channel, true);
+        public Channel JoinChannel(Channel channel, bool switchToJoined = false) => joinChannel(channel, switchToJoined, true);
```
I don't like this parameter existing, it doesn't need to exist, and it's also half-misleading (`JoinChannel` will actually switch to the joined channel *anyway* if `CurrentChannel.Value` is null). `ChatOverlay` could just as well set `CurrentChannel.Value` on its own without adding on weird boolean switches to pre-existing methods.

frenzibyte replied:
The point of this existing here was to switch the channel when it's populated with the messages and all (not immediately switch), but now that I look at it, fetching messages is done in a separate step which may take long before succeding, potentially after `HighlightedMessage` has been set and consumed... probably need to make `DrawableChannel` re-attempt highlighting on new messages rather than just forgetting about it.

bdach replied:
On re-reading I'm actually intensely confused by whatever it is that `joinChannel()` is doing. It appears to be recursively calling itself in some cases...? Very strange. But yeah as you said this flag you have set there doesn't do anything to protect against the fetch. To be robust you'd have to check for highlight in two places rather than rely on the bindable value change event.

bdach commented on osu.Game/Overlays/ChatOverlay.cs:
```diff
@@ -303,6 +303,29 @@ private void currentChannelChanged(ValueChangedEvent<Channel> e)
+                if (targetChannel != null)
+                    channelManager.CurrentChannel.Value = targetChannel;
+                else
+                    targetChannel = channelManager.JoinChannel(channelManager.AvailableChannels.SingleOrDefault(c => c.Id == message.ChannelId), true);
```
This can fail if the message's channel isn't found in `AvailableChannels` - `JoinChannel()` returns null if passed a null channel, and this will die on a nullref at the `HighlightedMessage` assignment.

I'd wager that this is going to fail when you open a PM channel, leave it, and then try to highlight the message in the now-closed PM channel, as PM channels are not in `AvailableChannels`.

frenzibyte replied:
Good catch, I should probably just pass the channel from the `MessageNotification` instead of doing this.

frenzibyte committed:
Revert "Add functionality to switch to successfully joined channel"

This reverts commit c72e8a8b5e37d3e959a491e926422ded94dd931a.

frenzibyte committed:
Improve channel switching flow in `HighlightMessage`

frenzibyte committed:
Improve message highlight handling in `DrawableChannel`

peppy commented on osu.Game/Overlays/Chat/DrawableChannel.cs:
```diff
@@ -79,6 +82,36 @@ private void load()
+        /// Processes any pending message in <see cref="highlightedMessage"/>.
+        /// </summary>
+        /// <remarks>
+        /// <see cref="CompositeDrawable.ScheduleAfterChildren"/> is for ensuring the scroll flow has updated with any new chat lines.
```
Weird to put this in xmldoc. It's an implementation detail so it should be an inline comment.

frenzibyte replied:
Hmm, I thought we usually leave implementation details in the remarks section. The sole reason I put it there is to benefit from referencing methods, but will move to comment if that's not where it should be placed.

peppy commented on osu.Game/Overlays/Chat/DrawableChannel.cs:
```diff
@@ -79,6 +82,36 @@ private void load()
+            base.LoadComplete();
+
+            highlightedMessage = Channel.HighlightedMessage.GetBoundCopy();
+            highlightedMessage.BindValueChanged(_ => processMessageHighlighting(), true);
```
```suggestion
            highlightedMessage.BindValueChanged(_ => Scheduler.AddOnce(processMessageHighlighting), true);
```

Probably?

frenzibyte replied:
The method is already scheduled, since all usages are not safe to call it directly. I guess it's better to schedule at usages rather than in the method itself?

peppy replied:
yeah i saw that. was more of a suggestion to avoid it running more than once, but i guess it's not the end of the world if it does.

peppy commented on osu.Game/Overlays/Chat/DrawableChannel.cs:
```diff
@@ -79,6 +82,36 @@ private void load()
+            scroll.ScrollTo(Math.Clamp(center, 0, scroll.ScrollableExtent));
+            chatLine.Highlight();
+
+            highlightedMessage.Value = null;
```
I guess we're still nulling this in the end?

frenzibyte replied:
Have mentioned one edge case of removing clearing this bindable in https://github.com/ppy/osu/pull/17137#discussion_r824005909.

frenzibyte committed:
Merge branch 'master' into chat-mention-highlight

frenzibyte committed:
Avoid running message highlight processing more than once

frenzibyte committed:
Move implementtaion detail to inline comment
